### PR TITLE
refactor!: improve support for ec2 instance workers

### PR DESCRIPTION
**Pull request by ddneilson**
### What was the problem/requirement? (What/Why)

The PosixInstanceWorker and WindowsInstanceWorker classes implement a CMF Worker based on an ec2 host. In both cases they start with an empty AMI image (AL2023 or Win Server 2022) then set up users, install the agent, etc. Tests that need to use an AMI that already has users and the agent installed end up having to hack around a lot of assumptions in these classes, or reimplement everything for an ec2-based worker from scratch.

### What was the solution? (How)

Refactor the PosixInstanceWorker and WindowsInstanceWorker to split them into os-specific base classes that contain all of the generic worker implementation stuff (starting the worker, getting its id, etc), and a derived class that implements the user-setup, agent install, and the like. This makes space for tests that use an AMI with the users & agent pre-setup to derive from the os-specific base classes to get the generic ec2-worker test functionality.

### What is the impact of this change?

Downstream consumers no longer need to resort to hacky workarounds to test worker configurations that have the agent & host setup baked into an AMI

### How was this change tested?

I've run this with the `deadline-cloud-worker-agent` e2e tests; the windows & linux e2e tests, and the cross-os tests under each.

I've also run this against internal testing infrastructure that is providing a pre-built AL2023 and WindowsServer2022 AMI to the testing infrastructure.

Everything passes everywhere. The `deadline-cloud-worker-agent` e2e tests need no modifications to take on these changes.

### Was this change documented?

No

### Is this a breaking change?

BREAKING CHANGE: PosixInstanceWorker has been renamed to PosixInstanceBuildWorker, and WindowsInstanceWorker has been renamed to WindowsInstanceBuildWorker.

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: src/deadline_test_fixtures/deadline/worker.py [hunk 1]
```diff
@@ -434,7 +436,7 @@ def _start_worker_agent(self) -> None:
                 [
                     "echo 'Running Get-Process to check if the agent is running'",
                     'for($i=1; $i -le 30 -and "" -ne $err ; $i++){sleep $i; Get-Process pythonservice -ErrorVariable err}',
-                    "IF(Get-Process pythonservice){echo 'service is running'}ELSE{exit 1}",
+                    "IF(Get-Process pythonservice){echo '+++SERVICE IS RUNNING+++'}ELSE{echo '+++SERVICE NOT RUNNING+++'; Get-Content -Encoding utf8 C:\ProgramData\Amazon\Deadline\Logs\worker-agent-bootstrap.log,C:\ProgramData\Amazon\Deadline\Logs\worker-agent.log; exit 1}",
```
Comment (ddneilson) [inline src/deadline_test_fixtures/deadline/worker.py:439]: I found failures in this script to be challenging to debug if the agent doesn't start. The reason was typically something discoverable in the agent logs, so I've added outputting the logs on error.

File: src/deadline_test_fixtures/deadline/worker.py [hunk 2]
```diff
@@ -443,40 +445,25 @@ def _start_worker_agent(self) -> None:
...
         self.worker_id = self.get_worker_id()
 
-    def configure_worker_command(self, *, config: DeadlineWorkerConfiguration) -> str:
-        """Get the command to configure the Worker. This must be run as root."""
+    def configure_worker_common(self, *, config: DeadlineWorkerConfiguration) -> str:
+        """Get the command to configure the Worker. This must be run as Administrator.
+        This cannot assume that the agent user exists.
+        """
 
-        cmds = [
```
Comment (ddneilson) [inline src/deadline_test_fixtures/deadline/worker.py:449]: All "install agent & setup agent user" stuff has been moved out of the base class and into the derived class.

File: src/deadline_test_fixtures/deadline/worker.py [hunk 3]
```diff
@@ -492,44 +479,6 @@ def configure_worker_command(self, *, config: DeadlineWorkerConfiguration) -> st
 
         return "; ".join(cmds)
 
-    def userdata(self, s3_files) -> str:
```
Comment (ddneilson) [inline src/deadline_test_fixtures/deadline/worker.py:495]: userdata is exclusively in the derived class

File: src/deadline_test_fixtures/deadline/worker.py [hunk 4]
```diff
@@ -543,14 +492,6 @@ def stop_worker_service(self):
 
         assert cmd_result.exit_code == 0, f"Failed to stop Worker Agent service: : {cmd_result}"
 
-    def ami_ssm_param_name(self) -> str:
```
Comment (ddneilson) [inline src/deadline_test_fixtures/deadline/worker.py:546]: This is exclusively in the derived class now.

File: src/deadline_test_fixtures/deadline/worker.py [hunk 5]
```diff
@@ -571,13 +512,99 @@ def get_worker_id(self) -> str:
...
+    Any commands must be written in Powershell.
     """
 
-    AL2023_AMI_NAME: ClassVar[str] = "al2023-ami-kernel-6.1-x86_64"
+    WIN2022_AMI_NAME: ClassVar[str] = "Windows_Server-2022-English-Full-Base"
+
+    def configure_worker_command(self, *, config: DeadlineWorkerConfiguration) -> str:
+        """Get the command to configure the Worker. This must be run as Administrator."""
+        cmds = [
+            "Set-PSDebug -trace 1",
```
Comment (ddneilson) [inline src/deadline_test_fixtures/deadline/worker.py:526]: The debug tracing is an intentional addition. If this ssm command fails, then it's challenging to understand why without it.

File: src/deadline_test_fixtures/deadline/worker.py [hunk 6]
```diff
@@ -571,13 +512,99 @@ def get_worker_id(self) -> str:
...
+        ]
+
+        if config.service_model_path:
+            cmds.append(
+                f"Copy-Item -Path ~\\.aws\\* -Destination C:\\Users\\{config.agent_user}\\.aws\\models -Recurse; "
+            )
+
+        return "; ".join(cmds)
+
+    def userdata(self, s3_files) -> str:
```
Comment (ddneilson) [inline src/deadline_test_fixtures/deadline/worker.py:551]: No changes here. Just a copy-over.

File: src/deadline_test_fixtures/deadline/worker.py [hunk 7]
```diff
@@ -595,28 +625,112 @@ def _start_worker_agent(self) -> None:
...
+                            # Verify Worker is still running
+                            "echo Waiting 5s for agent to get started",
+                            "sleep 5",
+                            "echo 'Running pgrep to see if deadline-worker-agent is running'",
+                            # Note: pgrep has a non-zero exit code if no matching application can be found.
+                            f"pgrep --count --full -u {self.configuration.agent_user} deadline-worker-agent",
+                        ],
+                    ),
+                    # If the worker didn't start, then print out the agent logs that exist to aid in debugging.
+                    "if test $? -ne 0; then echo '+++AGENT NOT RUNNING+++'; cat /var/log/amazon/deadline/worker-agent-bootstrap.log /var/log/amazon/deadline/worker-agent.log; exit 1; fi",
```
Comment (ddneilson) [inline src/deadline_test_fixtures/deadline/worker.py:643]: Added to make it much easier to diagnose issues from the logs when the agent fails to start.

File: src/deadline_test_fixtures/deadline/worker.py (hunk 1 above)
Comment (AWS-Samuel) [reply src/deadline_test_fixtures/deadline/worker.py:439]: 👏 👏 👏 This is so handy

File: src/deadline_test_fixtures/deadline/worker.py [hunk 8]
```diff
@@ -595,28 +625,112 @@ def _start_worker_agent(self) -> None:
...
+
+        cmd_result = self.send_command("systemctl start deadline-worker")
+
+        assert cmd_result.exit_code == 0, f"Failed to start Worker Agent service: {cmd_result}"
+
+    def stop_worker_service(self):
+        LOG.info("Sending command to stop the Worker Agent service")
+        cmd_result = self.send_command("systemctl stop deadline-worker")
+
+        assert cmd_result.exit_code == 0, f"Failed to start Worker Agent service: {cmd_result}"
```
Comment (AWS-Samuel) [inline src/deadline_test_fixtures/deadline/worker.py]:
```suggestion
        assert cmd_result.exit_code == 0, f"Failed to stop Worker Agent service: {cmd_result}"
```

**Review by AWS-Samuel — APPROVED** [review]
One small nit

**Review by AWS-Samuel — DISMISSED** [review]

**Review by YutongLi291 — DISMISSED** [review]

**Review by AWS-Samuel — APPROVED** [review]

**Review by YutongLi291 — APPROVED** [review]

**Pull request merged**